packages/playwright-core/src/server/injected/injectedScript.ts

```diff
@@ -1029,7 +1029,7 @@ export class InjectedScript {
       } else if (expression === 'to.be.enabled') {
         elementState = progress.injectedScript.elementState(element, 'enabled');
       } else if (expression === 'to.be.focused') {
-        elementState = document.activeElement === element;
+        elementState = document.activeElement === element || document.activeElement?.shadowRoot?.activeElement === element;
```

Comment:
I don't think this contradicts our doc spec (https://playwright.dev/docs/test-assertions#locator-assertions-to-be-focused):

```
Ensures the [Locator](https://playwright.dev/docs/api/class-locator) points to a focused DOM node.
```

since I'd include _shadow_ DOM node in "focused DOM node".

Reply:
This sounds good, but I think we should copy implementation from: https://github.com/microsoft/playwright/blob/57fcb590f7384e50c05671a5d8467da1bfd5fd32/packages/playwright-core/src/server/injected/injectedScript.ts#L677-L678

Reply:
Updated. Thanks! I copied those two lines, then decided to put in separate function that can be shared. I don't like the latter much due to the need to return the `activeElement` for the first use-case, but fine with either. wdyt?